akka-http/server/src/main/scala/endpoints4s/akkahttp/server/Endpoints.scala

```diff
@@ -449,63 +410,45 @@ trait EndpointsWithCustomErrors
+        currentRequest.matchAndParseHeadersDirective.flatMap { validatedUrlAndHeaders =>
+          Directives.extractRequest.map(heads.decode).map { validatedHeader =>
+            validatedUrlAndHeaders.zip(validatedHeader)
+          }
+        }
 
-      def aggregateAndValidate(
-          urlData: UrlData,
-          entityData: EntityData,
-          headersData: HeadersData
-      ): Validated[tupler.Out] =
+      def parseEntityDirective(urlAndHeaders: UrlAndHeaders): Directive1[tupler.Out] =
         currentRequest
-          .aggregateAndValidate(urlData, entityData, headersData._1)
-          .map(tupler(_, headersData._2))
+          .parseEntityDirective(urlAndHeaders._1)
+          .map(a => tupler(a, urlAndHeaders._2))
 
-      private[server] lazy val matchAndParseHeadersDirective
-          : Directive1[Validated[(UrlData, HeadersData)]] =
-        matchAndProvideParsedUrlAndHeadersData(method, url, headers)
+      def uri(out: tupler.Out): Uri =
+        currentRequest.uri(tupler.unapply(out)._1)
     }
 
   override def addRequestQueryString[A, Q](
       request: Request[A],
       qs: QueryString[Q]
   )(implicit tupler: Tupler[A, Q]): Request[tupler.Out] =
     new Request[tupler.Out] {
-      type UrlData = (request.UrlData, Q)
-      type HeadersData = request.HeadersData
-      type EntityData = request.EntityData
-
-      def method: Method = request.method
-      def url: Url[UrlData] = request.url.addQueryString(qs)
-      def entity: RequestEntity[EntityData] = request.entity
-      def headers: RequestHeaders[HeadersData] = request.headers
-
-      def urlData(o: tupler.Out): UrlData = {
-        val (a, q) = tupler.unapply(o)
-        val urlData = request.urlData(a)
-        (urlData, q)
-      }
+      type UrlAndHeaders = (request.UrlAndHeaders, Q)
 
-      private[server] def aggregateAndValidate(
-          urlData: UrlData,
-          entityData: EntityData,
-          headersData: HeadersData
-      ): Validated[tupler.Out] =
-        request.aggregateAndValidate(urlData._1, entityData, headersData).map(tupler(_, urlData._2))
+      lazy val matchAndParseHeadersDirective: Directive1[Validated[UrlAndHeaders]] =
+        request.matchAndParseHeadersDirective.flatMap { validatedUrlAndHeaders =>
+          Directives.parameterMultiMap.map(qs.validate).map { validatedQuery =>
+            validatedUrlAndHeaders.zip(validatedQuery)
+          }
+        }
 
-      private[server] lazy val matchAndParseHeadersDirective
-          : Directive1[Validated[(UrlData, HeadersData)]] =
-        matchAndProvideParsedUrlAndHeadersData(method, url, headers)
+      def parseEntityDirective(urlAndHeaders: (request.UrlAndHeaders, Q)): Directive1[tupler.Out] =
+        request
+          .parseEntityDirective(urlAndHeaders._1)
+          .map(a => tupler(a, urlAndHeaders._2))
+
+      def uri(out: tupler.Out): Uri =
+        request.uri(tupler.unapply(out)._1) // TODO Add query parameter
```

Comment:
I suppose this TODO needs to be fixed before merging?

Reply:
Yes, I forgot that one. And there is no test that covers it.

Reply:
I juste pushed a fixup commit to address this point.